### PR TITLE
Bind expressions

### PR DESCRIPTION
**Pull request by choumx**
Partial for #6199.

Implements Bind's JS-like expression grammar with Jison (also used by amp-access) and unit tests. Contains several security features, including:
- Disallow access of prototype properties
- Only allow invocation of whitelisted `String` and `Array` functions
- Disallow access to globals, function declaration and non-whitelisted keywords

Other things to note:
- Undefined vars, properties, array-index-out-of-bounds return null instead of throwing errors (similar to Angular)
- `bind-expr-impl.js` is just a compilation of `bind-expr-impl.jison` and can be ignored

/to @dvoytenko @aghassemi
/cc @cramforce

### CODE REVIEW

**Comment by aghassemi** [comment]
Not forgotten, will be :eyes: at this today.

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+        if ($1 === null || $2 === null) {
+          return;
+        }
+
+        var obj = Object.prototype.toString.call($1);
+        var prop = Object.prototype.toString.call($2);
+
+        if (obj === '[object Array]') {
+          if (prop === '[object Number]' && Number.isInteger($2)) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: `Number.isInteger` is not available everywhere. An imperfect but practical equivalent is `$2 === ($2 | 0)`, or you can go for the full [polyfill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger#Polyfill).

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            if ($2 >= 0 && $2 < $1.length) {
+              $$ = $1[$2];
+            }
+          }
+        }
+
+        if (prop === '[object String]') {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Numeric props are valid for objects, too.

**Comment by cramforce** [comment]
Regarding @jridgewell's speed comment: How long does it take to execute 100 typical expressions on a Nexus 5X type device?

File: extensions/amp-bind/0.1/test/test-bind-expr.js [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+  it('should evaluate undefined vars and properties to null', () => {
+    expect(evaluateBindExpr('foo')).to.be.null;
+    expect(evaluateBindExpr('foo.bar')).to.be.null;
+    expect(evaluateBindExpr('foo["bar"]')).to.be.null;
+    expect(evaluateBindExpr('foo[bar]')).to.be.null;
+    expect(evaluateBindExpr('foo[0]')).to.be.null;
+  });
+
+  it('should NOT allow access to prototype properties', () => {
+    expect(evaluateBindExpr('{}.constructor')).to.be.null;
```
Comment (cramforce) [inline extensions/amp-bind/0.1/test/test-bind-expr.js]: Also as bare expression: `expect(evaluateBindExpr('constructor')).to.be.null;`?

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+          if (Object.prototype.hasOwnProperty.call($1, $2)) {
```
Comment (aghassemi) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: these three lines alone should be able handle all the cases, regardless of whether `obj` is an array or not. For array access, both `arr[0] and arr['0']` are valid. Also`arr.hasOwnProperty(0) and arr.hasOwnProperty('0')` return true.

File: extensions/amp-bind/0.1/test/test-bind-expr.js [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+  const argumentTypeError = 'Unexpected argument type';
+  const unsupportedFunctionError = 'not a supported function';
+
+  it('should evaluate arithmetic operations', () => {
+    expect(evaluateBindExpr('-1')).to.equal(-1);
+    expect(evaluateBindExpr('1 + 2')).to.equal(3);
+    expect(evaluateBindExpr('2 - 3.5')).to.equal(-1.5);
+    expect(evaluateBindExpr('3 * 4')).to.equal(12);
+    expect(evaluateBindExpr('4 / 5')).to.equal(0.8);
+    expect(evaluateBindExpr('5 % 4')).to.equal(1);
```
Comment (aghassemi) [inline extensions/amp-bind/0.1/test/test-bind-expr.js]: add test for divide by 0.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Removed this code block per Ali's comment below.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 2 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Done.

File: extensions/amp-bind/0.1/test/test-bind-expr.js (hunk 3 above)
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-expr.js]: Done.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 4 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Good point, thanks!

**Comment by choumx** [comment]
> How long does it take to execute 100 typical expressions on a Nexus 5X type device?

~100ms to evaluate 100 complex-ish expressions (8 operations, 1 function call, 2 member access) on Nexus 5X. About half that on iPhone 6S.

**Comment by cramforce** [comment]
Maybe we should design this to run in a Worker from day 1.

File: extensions/amp-bind/0.1/bind-expr.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {parser} from './bind-expr-impl';
+
+export function evaluateBindExpr(expr, data) {
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-expr.js]: I don't believe jison has an option to build ast

**Comment by jridgewell** [comment]
How much of that 100ms is just the lexer? Ie, if we take away all the security code you wrote, how long does it take?

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+      {$$ = [];}
+  | '[' array ']'
+      {$$ = $2;}
+  ;
+
+array:
+    expr
+      {$$ = [$1];}
+  | array ',' expr
+      {$$ = $1; Array.prototype.push.call($1, $3);}
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: `$1.push($3)`?

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+  ;
+
+variable:
+    NAME
+      {$$ = Object.prototype.hasOwnProperty.call(yy, $1) ? yy[$1] : null;}
+  ;
+
+literal:
+    STRING
+      {$$ = yytext.substr(1, yyleng - 2);}
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Why the substring?

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+member:
+    '.' NAME
+      {$$ = $2;}
+  | '[' expr ']'
+      {$$ = $2;}
+  ;
+
+variable:
+    NAME
+      {$$ = Object.prototype.hasOwnProperty.call(yy, $1) ? yy[$1] : null;}
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Since `yy` can be `null`, we'll have to guard that.

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+      String.prototype.toLowerCase,
+      String.prototype.toUpperCase,
+    ],
+};
+
+/** @return {bool} Returns false if args contains an invalid type. */
+function typeCheckArgs(args) {
+  for (var i = 0; i < args.length; i++) {
+    var arg = args[i];
+    if (Object.prototype.toString.call(arg) === '[object Object]') {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: `typeof arg === "object" && !Array.isArray(arg);`

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+  ;
+
+invocation:
+    expr '.' NAME args
+      %{
+        $$ = null;
+
+        var obj = Object.prototype.toString.call($1);
+
+        var whitelist = functionWhitelist[obj];
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Instead of `toString`, use an if:

```js
var whitelist;
if (typeof obj === "string") {
} else if (Array.isArray(obj)) {
}

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        if (whitelist) {
+          var fn = $1[$3];
+          if (whitelist.indexOf(fn) >= 0) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Since they're method on the prototype, names can't conflict. That makes it perfect for an prototype-less object map instead of an array.

```js
whitelist[$3] === $1[$3];

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+            if (typeCheckArgs($4)) {
+              $$ = fn.apply($1, $4);
+            } else {
+              throw new Error(`Unexpected argument type in {$3}()`);
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: This throw should be in `typeCheckArgs()`

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    expr member
+      %{
+        $$ = null;
+
+        if ($1 === null || $2 === null) {
+          return;
+        }
+
+        var prop = Object.prototype.toString.call($2);
+        if (prop === '[object String]' || prop === '[object Number]') {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: These type checks are likely unnecessary. We definitely need to enforce that it is **not** a `Symbol`, but we can do that much quicker at a higher level (forbid the key in the `yy` context object).

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+          if (Object.prototype.hasOwnProperty.call($1, $2)) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Hoist `Object#hasOwnProperty` into a variable to avoid repeated lookups.

File: extensions/amp-bind/0.1/test/test-bind-expr.js [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+    expect(evaluateBindExpr('Intl')).to.be.null;
+
+    expect(evaluateBindExpr('Iterator')).to.be.null;
+    expect(evaluateBindExpr('ParallelArray')).to.be.null;
+    expect(evaluateBindExpr('StopIteration')).to.be.null;
+  });
+
+  /** @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators */
+  it('should NOT allow access to non-whitelisted operators', () => {
+    expect(evaluateBindExpr('this')).to.be.null;
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/test/test-bind-expr.js]: let's also add `self` and `global`

File: extensions/amp-bind/0.1/bind-expr.js (hunk 6 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr.js]: Jison doesn't support it natively, but I could build a custom one in the grammar. Not sure how much of a speed-up we'd get but I'll try it out.

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+      {$$ = $1 < $3;}
+  | expr '>=' expr
+      {$$ = $1 >= $3;}
+  | expr '>' expr
+      {$$ = $1 > $3;}
+  | expr '!=' expr
+      {$$ = $1 != $3;}
+  | expr '==' expr
+      {$$ = $1 == $3;}
+  | expr '?' expr ':' expr
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Is there elvis too?

File: extensions/amp-bind/0.1/test/test-bind-expr.js [hunk 18]
```diff
@@ -0,0 +1,430 @@
...
+    expect(() => { evaluateBindExpr('foo--', {foo: 0}); }).to.throw();
+    expect(() => { evaluateBindExpr('~foo', {foo: 0}); }).to.throw();
+    expect(() => { evaluateBindExpr('foo << 1', {foo: 0}); }).to.throw();
+    expect(() => { evaluateBindExpr('foo >> 1', {foo: 0}); }).to.throw();
+    expect(() => { evaluateBindExpr('foo >>> 1', {foo: 0}); }).to.throw();
+    expect(() => { evaluateBindExpr('new Object()', {foo: 0}); }).to.throw();
+    expect(() => { evaluateBindExpr('delete foo', {foo: 0}); }).to.throw();
+  });
+
+  it('should NOT allow control flow or loops', () => {
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/test/test-bind-expr.js]: Not seeing tests for `?:`, but they seem to be allowed?

**Comment by dvoytenko** [comment]
@choumx Wow! This grammar is way more powerful than I expected. But, I guess, why hide from it :)

**Comment by choumx** [comment]
> How much of that 100ms is just the lexer? Ie, if we take away all the security code you wrote, how long does it take?

The security code seems to be only a small fraction (< 5%) of the total runtime. I'm going to experiment with building an AST first -- hopefully that will boost performance substantially.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 15 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Done.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 12 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Nice, this will avoid scanning the whitelist array. Done.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 9 above)
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: It should throw an error.

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 19]
```diff
@@ -1,41 +1,45 @@
...
   '[object Array]':
-    [
-      Array.prototype.concat,
-      Array.prototype.includes,
-      Array.prototype.indexOf,
-      Array.prototype.join,
-      Array.prototype.lastIndexOf,
-      Array.prototype.slice,
-    ],
+    {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: These must be prototype-less objects. As is, you've just allowed `__defineGetter__`, `__defineSetter__`, `hasOwnProperty`, `isPrototypeOf`, etc.

**Comment by cramforce** [comment]
Good for us we have Justin on our side :) Lets also test them.

**Comment by cramforce** [comment]
types.js#map can help with creating such an object from a literal.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 19 above)
Comment (cramforce) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Good for us we have Justin on our side :) Lets also test them.

`types.js#map` can help with creating such an object from a literal.

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 9 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Actually we want to return `null` instead of throwing undefined symbol error to be more forgiving (see PR summary).

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 14 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: I actually had to revert this change since `hasOwnProperty` throw a "cannot convert to primitive" error in some cases. Changed to use `typeof` instead of `toString` though, which should be faster.

**Comment by choumx** [comment]
Re: performance, I looked around and there are other parser generator options. This [comparison](https://sap.github.io/chevrotain/performance/) suggests that we may get several-X speed up with another generator or by handwriting a parser. @jridgewell also suggested precompiling the AST into a Function (underscore and Angular do this).

I'll continue to investigate expression performance as I work on an end-to-end MVP for Bind.

**Comment by cramforce** [comment]
Lets get this in and then optimize. We can chat offline about details, but I'm pretty bullish about moving all of this into a web worker.

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 20]
```diff
@@ -2,44 +2,57 @@
...
+  // Creates a prototype-less map of function name to the function itself.
+  // This makes function lookups faster (compared to Array.indexOf).
+  const out = Object.create(null);
+  Object.keys(whitelist).forEach(type => {
+    out[type] = Object.create(null);
+
+    const functions = whitelist[type];
+    for (let i = 0; i < functions.length; i++) {
+      const f = functions[i];
+      out[type][f.name] = f;
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: The name property has spotty support in iOS and Android Browser (non chrome). Instead of that, we could use strings and grab the method off the prototype:

```js
const whitelist = [
  {
    class: '[object Array]',
    proto: Array.prototype,
    methods: ['concat', 'includes', ...]
  },
];

File: extensions/amp-bind/0.1/bind-expr-impl.jison [hunk 21]
```diff
@@ -2,44 +2,57 @@
 
 %{
 // Shortcuts for common functions.
-var toString = Object.prototype.toString;
-var hasOwnProperty = Object.prototype.hasOwnProperty;
+const toString = Object.prototype.toString;
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Is this compiled into es5?

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 20 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: For this simple usage of `Function.name`, it appears to be [supported](http://kangax.github.io/compat-table/es6/#test-function_name_property).

File: extensions/amp-bind/0.1/bind-expr-impl.jison (hunk 17 above)
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-expr-impl.jison]: Nope. `||` is good enough.

File: extensions/amp-bind/0.1/test/test-bind-expr.js [hunk 22]
```diff
@@ -0,0 +1,430 @@
...
+    expect(evaluateBindExpr('true && false || true')).to.be.true;
+    expect(evaluateBindExpr('true && false == false')).to.be.true;
+    expect(evaluateBindExpr('false || false == true')).to.be.false;
+    expect(evaluateBindExpr('false == !true')).to.be.true;
+  });
+
+  it('should support strings', () => {
+    expect(evaluateBindExpr('"a"')).to.equal('a');
+    expect(evaluateBindExpr('"a".length')).to.equal(1);
+    expect(evaluateBindExpr('"a" + "b"')).to.equal('ab');
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/test/test-bind-expr.js]: I'm ok deferring. But we need to see the most prevalent cases while we are implementing it toward MVP. If it's URL construction, I believe, we need to do it for MVP.

**Comment by dvoytenko** [comment]
LGTM with one request for an additional test.

**Comment by choumx** [comment]
Thanks for the reviews everyone!
